### PR TITLE
backend authentication by user role

### LINKED ISSUE #71: user roles
- set a user role api or middleware when there is a token, the user roles who are just normal user cannot access admin routes
-  we can set this up as a middle ware or api

### PR DESCRIPTION
**Pull request by fidel-bello**
# Pull request

## Description

[//]: # (Summary of changes. List any dependency changes required for this change)
image
if a user role tries to access admin, it will cause a fail or error

closes #71 
## Type of Change

- [ ] Refactor (non-breaking change which improves code performance or readability)
- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change that adds functionality)
- [ ] Documentation (changes to docs and not code)
- [ ] Breaking Change (Fix or change that would cause existing functionality to not work as expected)
- [ ] Test

## How it was tested

Enter description of testing

- [ ] Unit Test
- [x] Manual / Postman(if backend)
- [ ] Integration

## Affirmations

- [x] I follow coding guidelines (TBD)
- [x] I have self-reviewed my code
- [ ] I have run eslint and fixes issues (Mostly works but needs more rules and a fix)
- [ ] I have commented code when necessary or difficult
- [ ] I have made changes to docs where required
- [x] My changes generate no new warnings
- [x] I have added tests
- [x] New and existing tests pass

### CODE REVIEW

File: Backend/src/app/models/user.ts [hunk 1]
```diff
@@ -31,11 +31,11 @@ const userSchema = new Schema({
     },
     role: {
         type: String,
-        default: 'user' //defaults to normal user
+        default: 'user'  //defaults to normal user
     },
     createdAt: {
         type: Date,
-        default: Date.now
+        default: Date.now()
```
Comment (ivan-ngchakming) [inline Backend/src/app/models/user.ts:38]: Are you sure? In my experience, its usually the function that gets passed as default.

**Review by fidel-bello — COMMENTED** [review]
yes.. according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now that is the correct syntax. but it works either way. im just going based off the docs.

File: Backend/src/app/models/user.ts (hunk 1 above)
Comment (nmp14) [reply Backend/src/app/models/user.ts:38]: Date.now will save a function that returns the time on a user creation.
Date.now() will return the time of schema creation as it gets called immediately. Probably not ideal.

Comment (fidel-bello) [reply Backend/src/app/models/user.ts:38]: oh I see now, thanks for the explanation will change now.

Comment (ivan-ngchakming) [reply Backend/src/app/models/user.ts:38]: > yes.. according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now that is the correct syntax. but it works either way. im just going based off the docs.

@fidel-bello It doesn't work, note the creation timestamps of each user

image

File: Backend/src/app/controllers/authControllers.ts [hunk 2]
```diff
@@ -4,11 +4,12 @@ import { Error_Handler } from "../utils/errorHandling";
 import { User } from "../models/user";
 import jwt from "jsonwebtoken";
 import { sendToken } from "../middlewares/jwtToken";
+import process from "process";
```
Comment (ivan-ngchakming) [inline Backend/src/app/controllers/authControllers.ts:7]: Why is this import required?

Comment (ivan-ngchakming) [inline Backend/src/app/controllers/authControllers.ts:7]: Why is this import required?

File: Backend/src/app/controllers/authControllers.ts [hunk 3]
```diff
@@ -69,5 +70,12 @@ export class Auth_Controllers
             success: true,
             message: "Logged out"
         })
-    }) 
+    })
+
+    public authorizeRoles = (...roles: any[]) =>  {
```
Comment (ivan-ngchakming) [inline Backend/src/app/controllers/authControllers.ts:75]: I don't think it is a good practice to use any, weird eslint/ts didnt catch this.

```suggestion
    public authorizeRoles = (...roles: Role[]) =>  {
```

And create a Role enum:
```js
    enum Role {
        Admin = 'admin',
        ...
    }
```

File: Backend/src/app/models/user.ts [hunk 4]
```diff
@@ -31,7 +31,7 @@ const userSchema = new Schema({
     },
     role: {
         type: String,
-        default: 'user' //defaults to normal user
+        default: 'user'  //defaults to normal user
```
Comment (ivan-ngchakming) [inline Backend/src/app/models/user.ts:34]: Use the Role enum mentioned above

File: Backend/src/app/routes/productRoutes.ts [hunk 5]
```diff
@@ -10,9 +10,9 @@ router.route('/products/:id').get(Product.getSingleProduct);
 
 //admin routes
 router.route('/admin/products/:id')
-    .put(User.isAuthenticated , Product.updateProduct)
-    .delete(User.isAuthenticated, Product.deleteProduct); //could be wrong format? did this because the admin product id route is used twice
+    .put(User.isAuthenticated, User.authorizeRoles('admin') , Product.updateProduct)
+    .delete(User.isAuthenticated, User.authorizeRoles('admin'), Product.deleteProduct); //could be wrong format? did this because the admin product id route is used twice
 
-router.route('/admin/products/new').post(User.isAuthenticated, Product.createProducts);
+router.route('/admin/products/new').post(User.isAuthenticated, User.authorizeRoles('admin'), Product.createProducts);
```
Comment (ivan-ngchakming) [inline Backend/src/app/routes/productRoutes.ts:16]: Use the Role enum mentioned above

**Review by ivan-ngchakming — CHANGES_REQUESTED** [review]
Looks good in general, just a minor typing issue

File: Backend/src/app/controllers/authControllers.ts (hunk 2 above)
Comment (fidel-bello) [reply Backend/src/app/controllers/authControllers.ts:7]: its not.. i was trying to do something with process.env and it automatically imported.. sorry didnt mean to

Comment (ivan-ngchakming) [reply Backend/src/app/controllers/authControllers.ts:7]: hey don't be sorry, gosh knows how many time I've committed mistakes like this to a PR. vscode should be the one who's sorry for not being smart enough

File: Backend/src/app/controllers/authControllers.ts [hunk 6]
```diff
@@ -72,10 +88,10 @@ export class Auth_Controllers
...
     })
 
-    public authorizeRoles = (...roles: any[]) =>  {
-        return (req: any, res:Response, next: NextFunction) => {
-            if(!roles.includes(req.user.role))
-               return next(new Error_Handler(`Role: "${req.user.role}" is not allowed to access this route`, 403));
+    public authorizeRoles = (...roles: UserRole[]) =>  {
+        return (req: Request, res:Response, next: NextFunction) => {
+            if(!roles.includes(UserRole.admin))
+               return next(new Error_Handler(`Role: "${UserRole.admin}" is not allowed to access this route`, 403));
```
Comment (ivan-ngchakming) [inline Backend/src/app/controllers/authControllers.ts:94]:
```suggestion
            if(!roles.includes(req.user.role))
               return next(new Error_Handler(`Role: "${req.user.role}" is not allowed to access this route`, 403));
```

File: Backend/src/app/controllers/authControllers.ts [hunk 7]
```diff
@@ -4,14 +4,30 @@ import { Error_Handler } from "../utils/errorHandling";
...
 import jwt from "jsonwebtoken";
 import { sendToken } from "../middlewares/jwtToken";
-import process from "process";
 
 
-export class Auth_Controllers
+enum UserRole
 {
-    public registerUser = asyncError( async (req: Request, res: Response, _next: NextFunction): Promise<void> => {
+    admin = 'admin'
```
Comment (ivan-ngchakming) [inline Backend/src/app/controllers/authControllers.ts:11]: Missed the default user role?

```suggestion
    Admin = 'admin',
    User = 'user',
```

**Review by ivan-ngchakming — DISMISSED** [review]
Looks good, just a few final suggestions

File: Backend/src/app/models/user.ts (hunk 4 above)
Comment (ivan-ngchakming) [inline Backend/src/app/models/user.ts:34]: My bad, missed one comment from last review

```suggestion
        default: UserRole.User  //defaults to normal user
```

**Pull request merged**